core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.Consumer;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A dynamic {@link EndpointGroup}. The list of {@link Endpoint}s can be updated dynamically.
+ */
+public class DynamicEndpointGroup implements EndpointGroup {
+    private final Set<Consumer<Collection<Endpoint>>> updateListeners = new CopyOnWriteArraySet<>();
+    private volatile List<Endpoint> endpoints = new CopyOnWriteArrayList<>();
+
+    @Override
+    public final List<Endpoint> endpoints() {
+        return endpoints;
+    }
+
+    /**
+     * Adds the specified {@link Endpoint} to current {@link Endpoint} list.
+     */
+    protected final void addEndpoint(Endpoint e) {
+        endpoints.add(e);
+        notifyListeners();
+    }
+
+    /**
+     * Removes the specified {@link Endpoint} from current {@link Endpoint} list.
+     */
+    protected final void removeEndpoint(Endpoint e) {
+        endpoints.remove(e);
+        notifyListeners();
+    }
+
+    /**
+     * Sets the specified {@link Endpoint}s as current {@link Endpoint} list.
+     */
+    protected final void setEndpoints(Iterable<Endpoint> endpoints) {
+        this.endpoints = new CopyOnWriteArrayList<>(ImmutableList.copyOf(endpoints));
```

Comment:
I realized that this has a chance of the race condition where `add/removeEndpoint()` writes on the `CopyOnWriteArrayList` that's been replaced already. How about reverting to the previous implementation and using a `ReentrantLock` explicitly on `add/remove/setEndpoint(s)`?